crates/ide-completion/src/item.rs

```diff
@@ -164,6 +164,8 @@ pub struct CompletionRelevance {
     pub postfix_match: Option<CompletionRelevancePostfixMatch>,
     /// This is set for type inference results
     pub is_definite: bool,
+    /// Any other bonuses we want to add,
+    pub bonus_score: u32,
```

Comment:
This defeats the purpose of `CompletionRelevance`, the whole point of it is to have fields representing properties that influence the final score. This field just spreads this final score across multiple locations. This should instead be an enum like

```rs
enum CompletionRelevanceConstructorType {
    /// Not a constructor
    None,
    /// Returns the Self type of the impl/trait
    DirectConstructor,
    /// Returns something that indirectly constructos the `Self` type of the impl/trait e.g. `Result<Self, ()>`, `Option<Self>`
    Constructor,
    /// Returns a possible builder for the type
    Builder,
}
```
and then assign points respectively where we calculate all the other points. Those variants are the interesting cases I can think of at least, `Constructor` might be tricky to implement though, but that can be skipped for now in that case.

Reply:
Thanks for some useful pointers, I've pushed a few commits.

> `Constructor` might be tricky to implement though, but that can be skipped for now in that case.

This needed some ugly string comparison - if this is not good enough then I can create a separate PR and attempt to implement it properly later.